### PR TITLE
Pinnable cancellations

### PR DESCRIPTION
**Pull request by syntheticmagus**
Changes the behavior of cancellation by making cancellation tokens "pin" the cancellation source under certain circumstances. Most importantly, cancellations get pinned while the work payload they are made to cancel is being executed. This makes cancelling a cancel source a blocking operation; a pinned cancellation source will block on cancel until it is unpinned. At the same time, cancelling *immediately* prevents any further pinning. The end result is that calling `cancel()` now means that no new work guarded by this cancellation can be started, but all work guarded by this cancellation will be finished before `cancel()` returns. This can be used, for example, to ensure that task can safely utilize resources without taking ownership of them. Suppose, for example, that a class `Foo` has a member variable `m_bar` and a cancellation source `m_cancelSource`. So long as `~Foo()` calls `m_cancelSource.cancel()` before it destructs `m_bar`, tasks run using `m_cancelSource` can safely use `m_bar` without worrying that it might be destructed by `Foo`'s lifecycle after the task has started but before the task completes.

### CODE REVIEW

File: Source/Shared/arcana/threading/task.h [hunk 1]
```diff
@@ -271,6 +280,8 @@ namespace arcana
         using traits = internal::callable_traits<CallableT, void>;
         using wrapper = internal::input_output_wrapper<void, typename traits::error_propagation_type, false>;
 
+        auto cancel_pin = token.pin();
```
Comment (bghgary) [inline Source/Shared/arcana/threading/task.h:283]: Maybe just a simple comment saying that taking the pin early is safer? Something like that.

File: Source/Shared/arcana/threading/task.h [hunk 2]
```diff
@@ -111,9 +111,18 @@ namespace arcana
                     })
             ) };
 
-            m_payload->create_continuation([&scheduler](auto&& c)
+            m_payload->create_continuation([this, &scheduler, token](auto&& c) mutable
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Source/Shared/arcana/threading/task.h:114]: Spoke about this fairly extensively offline. `this` is probably not a good thing to capture, but there are type erasure problems with passing it as an argument, so an alternative is to capture `m_payload` directly. `token` does, I think, need to be captured here because (1) it's guarding the scheduler, which is captured here, and (2) the `continuation_payload` which calls the scheduling lambda doesn't know about cancellation, so it doesn't have it to pass in.

File: Source/Shared/arcana/threading/cancellation.h [hunk 3]
```diff
@@ -52,16 +52,16 @@ namespace arcana
...
                 }
 
                 // We want to signal cancellation in reverse order
                 // so that if a parent function adds a listener
                 // then a child function does the same, the child
                 // cancellation runs first. This avoids ownership
                 // semantic issues.
-                for(auto itr = listeners.rbegin(); itr != listeners.rend(); ++itr)
+                for (auto itr = listeners.rbegin(); itr != listeners.rend(); ++itr)
                     (*itr)();
```
Comment (bghgary) [inline Source/Shared/arcana/threading/cancellation.h:65]: nit:
```suggestion
                for (auto itr = listeners.rbegin(); itr != listeners.rend(); ++itr) {
                    (*itr)();
                }
```

File: Source/Shared/arcana/threading/task.h [hunk 4]
```diff
@@ -120,8 +120,13 @@ namespace arcana
...
                 {
-                    if (!m_payload->completed())
-                        m_payload->complete({ make_unexpected(std::errc::operation_canceled) });
+                    // By the time we get here, payload is actually owned by the lambda in c.
+                    // However, due to type erasure, the continuation_payload calling this scheduling
+                    // lambda no longer knows the type information needed to correctly complete
+                    // the payload, so it must be independently captured here so that it can be
+                    // correctly completed in the cancellation case.
+                    if (!payload->completed())
+                        payload->complete({ make_unexpected(std::errc::operation_canceled) });
```
Comment (bghgary) [inline Source/Shared/arcana/threading/task.h:129]: nit:
```suggestion
                    if (!payload->completed()) {
                        payload->complete({ make_unexpected(std::errc::operation_canceled) });
                    }
```

File: Source/Shared/arcana/threading/cancellation.h [hunk 5]
```diff
@@ -2,28 +2,118 @@
...
+            using collection = ticketed_collection<std::function<void()>>;
+
+        public:
+            using ticket = collection::ticket;
+            using ticket_scope = collection::ticket_scope;
+
+            template<typename CallableT>
+            ticket add_listener(CallableT&& callback, std::function<void()>& copied)
+            {
+                std::scoped_lock<std::mutex> lock{ m_mutex };
```
Comment (ryantrem) [inline Source/Shared/arcana/threading/cancellation.h:30]: Why this change? Isn't scoped_lock heavier than lock_guard, and intended for cases where you are trying to lock on multiple mutexes?

**Review by ryantrem — COMMENTED** [review]
So with this change, it is not possible to request an async operation be cancelled without blocking the caller? Am I understanding this correctly?

File: Source/Shared/arcana/threading/cancellation.h (hunk 5 above)
Comment (bghgary) [reply Source/Shared/arcana/threading/cancellation.h:30]: I'm not aware that scoped_lock is heavier than lock_guard. The only reasoning for using lock_guard I could find is here: https://stackoverflow.com/questions/43019598/stdlock-guard-or-stdscoped-lock?msclkid=6aad9a81bc4111eca785163544f6d052

File: Source/Shared/arcana/threading/cancellation.h [hunk 6]
```diff
@@ -28,22 +28,17 @@ namespace arcana
...
-                {
-                    copied = std::forward<CallableT>(callback);
-                    return m_listeners.insert(copied, m_mutex);
-                }
-                else
-                {
-                    return m_listeners.insert(std::forward<CallableT>(callback), m_mutex);
-                }
+            template<typename CallableT>
+            ticket add_cancellation_completed_listener(CallableT&& callback, std::function<void()>& copied)
```
Comment (bghgary) [inline Source/Shared/arcana/threading/cancellation.h:35]: I think maybe just this without `cancellation` since this is the cancellation class already.
```suggestion
            ticket add_completed_listener(CallableT&& callback, std::function<void()>& copied)
```

File: Source/Shared/arcana/threading/cancellation.h [hunk 7]
```diff
@@ -203,20 +233,24 @@ namespace arcana
             return { m_impl };
         }
 
-        void unsafe_cancel()
+        void cancel(bool block_until_cancellation_completed = false)
```
Comment (bghgary) [inline Source/Shared/arcana/threading/cancellation.h:236]: nit: we seem to be using camelCase for variables and also remove `cancellation` since it's implied
```suggestion
        void cancel(bool blockUntilCompleted = false)
```

**Review by bghgary — APPROVED** [review]
